Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,149 @@
+    m_runningPositions = Vector<LayoutUnit>(m_gridAxisTracksCount);
+
+    // 2.4 Masonry Layout Algorithm
+    addItemsToFirstTrack(firstTrackItems);
+    placeItemsWithDefiniteGridAxisPosition(itemsWithDefiniteGridAxisPosition);
+    placeItemsWithIndefiniteGridAxisPosition(itemsWithIndefinitePosition);
+}
+
+void GridMasonryLayout::addItemsToFirstTrack(HashMap<RenderBox*, GridArea>& firstTrackItems)
+{
+    for (auto& itemItr : firstTrackItems) {
+        auto* item = itemItr.key;
+        auto gridArea = itemItr.value;
+
+        insertIntoGridAndLayoutItem(item, gridArea);
+
+        updateRunningPositions(item);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition(Vector<RenderBox*> itemsWithDefiniteGridAxisPosition)
+{
+    for (auto* item : itemsWithDefiniteGridAxisPosition) {
+        auto itemSpan = m_renderGrid.currentGrid().gridItemSpan(*item, m_gridAxisDirection);
+
+        ASSERT(!itemSpan.isIndefinite());
+
+        auto initialTrackSpan = GridSpan::translatedDefiniteGridSpan(0, 1);
+        itemSpan.translate(m_renderGrid.currentGrid().explicitGridStart(m_gridAxisDirection));
+
+        GridArea area = m_masonryAxisDirection == ForRows ? GridArea(initialTrackSpan, itemSpan) : GridArea(itemSpan, initialTrackSpan);
+        insertIntoGridAndLayoutItem(item, area);
+
+        updateRunningPositions(item);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition(Vector<RenderBox*> itemsWithIndefinitePosition)
+{
+    for (auto* item : itemsWithIndefinitePosition) {
+        insertIntoGridAndLayoutItem(item, nextMasonryPositionForItem(item));
+
+        updateRunningPositions(item);
+    }
+}
+
+void GridMasonryLayout::insertIntoGridAndLayoutItem(RenderBox* child, GridArea area)
+{
+    m_renderGrid.currentGrid().insert(*child, area);
+
+    if (m_gridAxisDirection == ForColumns)
+        m_renderGrid.updateGridAreaLogicalSize(*child, m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForColumns), std::nullopt);
+    else
+        m_renderGrid.updateGridAreaLogicalSize(*child, std::nullopt, m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
+    child->layoutIfNeeded();
+}
+
+void GridMasonryLayout::updateRunningPositions(RenderBox* child)
+{
+    auto gridSpan = m_masonryAxisDirection == ForRows ? m_renderGrid.currentGrid().gridItemArea(*child).columns : m_renderGrid.currentGrid().gridItemArea(*child).rows;
+
+    ASSERT(gridSpan.startLine() < m_runningPositions.size() && gridSpan.endLine() <= m_runningPositions.size());
+    gridSpan.clamp(m_runningPositions.size());
+
+    auto previousRunningPosition = m_runningPositions[gridSpan.startLine()];
+
+    auto newRunningPosition = m_masonryAxisDirection == ForRows ? previousRunningPosition + child->logicalHeight() + child->verticalMarginExtent(): previousRunningPosition + child->logicalWidth() + child->horizontalMarginExtent();
```

Comment:
Typically there is a space on both sides of the colon.

Reply:
Fixed